pkg/querier/remote_read.go

```diff
@@ -147,14 +147,72 @@ func remoteReadStreamedXORChunks(
 	}
 	w.Header().Set("Content-Type", api.ContentTypeRemoteReadStreamedChunks)
 
+	// Process all queries concurrently and collect their ChunkSeriesSet results
+	type queryResult struct {
+		idx    int
+		series storage.ChunkSeriesSet
+		err    error
+	}
+
+	resultCh := make(chan queryResult, len(req.Queries))
+
+	// Launch goroutines to process each query concurrently
 	for i, qr := range req.Queries {
-		if err := processReadStreamedQueryRequest(ctx, i, qr, q, w, f, maxBytesInFrame); err != nil {
+		go func(i int, qr *prompb.Query) {
+			start, end, _, _, matchers, hints, err := queryFromRemoteReadQuery(qr)
+			if err != nil {
+				resultCh <- queryResult{idx: i, err: err}
+				return
+			}
+
+			querier, err := q.ChunkQuerier(int64(start), int64(end))
+			if err != nil {
+				resultCh <- queryResult{idx: i, err: err}
+				return
+			}
+
+			// The streaming API has to provide the series sorted.
+			seriesSet := querier.Select(ctx, true, hints, matchers...)
+			resultCh <- queryResult{idx: i, series: seriesSet, err: nil}
+		}(i, qr)
+	}
+
+	// Collect results and maintain order
```

Comment:
Should we have some limits on the number of queries evaluated in parallel, or the amount of data they can return? Otherwise we could be holding a lot of data in memory here and potentially OOM.

Reply:
for our use in GEM, we need all queries to run concurrently. I can see where you're coming from. Is it ok if I open a follow-up PR to add concurrency controls?

Reply:
>Is it ok if I open a follow-up PR to add concurrency controls?

Sounds good to me.

Reply:
Just for my own understanding, is there a reason why we don't use `concurrency.ForEachJob` (from dskit) here? It would make it easier to implement concurrency controls, and cancellation of all goroutines if any fail, plus it's easier to maintain code if we reuse more unified methods wherever possible.